### PR TITLE
Introduce a new set of time APIs

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/1058

## Approach
Introduce support for:
- UTC time
- Civil time
- Other APIs to do time manipulations and conversions

## Release note
Introduce support for:
- UTC time
- Civil time
- Other APIs to do time manipulations and conversions

## Automation tests
 - Unit tests 
   > Yes

## Test environment
- macOS
- Java 11
- SL 3
 
## Learning
- https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/proposals/timestamp/timestamp.md
- https://tools.ietf.org/html/rfc8536
- https://www.python.org/dev/peps/pep-0495/

### CODE REVIEW

File: time-ballerina/time_apis.bal [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+# Returns Utc representing current time.
+# + precision - Specifies number of zeros after decimal point (e.g. 3 would give millisecond precision
+# and nil means native precision(nanosecond precision 9) of clock)
+# + return - The `time:Utc` value corresponding to the current UTC time
+public isolated function utcNow(int? precision = ()) returns Utc {
+    int precisionValue = -1;
+    if (precision is int) {
+        precisionValue = precision;
+    }
+    [int, Seconds] currentUtc = externUtcNow(precisionValue);
```
Comment (jclark) [inline time-ballerina/time_apis.bal:28]: Why doesn't externUtcNow return a readonly value?

File: time-ballerina/time_apis.bal [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    return <Utc>currentUtc.cloneReadOnly();
```
Comment (jclark) [inline time-ballerina/time_apis.bal:29]: Should not need to cast here.

Comment (jclark) [reply time-ballerina/time_apis.bal:29]: Why not just?

```
return externUtcNow(precision ?: -1).cloneReadonly();
```

File: time-ballerina/time_apis.bal [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+public isolated function monotonicNow() returns Seconds {
+    return externMonotonicNow();
+}
+
+# Converts from RFC 3339 timestamp(e.g. `2007-12-03T10:15:30.00Z`) to Utc.
+# + timestamp - RFC 3339 timestamp(e.g. `2007-12-03T10:15:30.00Z`) value as a string
+# + return - The corresponding `time:Utc` or a `time:Error` when the specified timestamp
+# is not adhere to the RFC 3339 format(e.g. `2007-12-03T10:15:30.00Z`)
+public isolated function utcFromString(string timestamp) returns Utc|Error {
+    [int, Seconds]|Error utc = externUtcFromString(timestamp);
```
Comment (jclark) [inline time-ballerina/time_apis.bal:43]:
```
return (check externUtcFromString(timestamp)).cloneReadOnly();
```

File: time-ballerina/time_apis.bal [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+public isolated function utcToString(Utc utc) returns string {
+    return externUtcToString(utc);
+}
+
+# Returns Utc time that occurs seconds after `utc`. This assumes that all days have 86400 seconds.
+# + utc - Utc time as a tuple `[int, decimal]`
+# + seconds - Number of seconds to be added
+# + return - The resulted `time:Utc` value after the summation
+public isolated function utcAddSeconds(Utc utc, Seconds seconds) returns Utc {
+    int secondsFromEpoch = utc[0];
```
Comment (jclark) [inline time-ballerina/time_apis.bal:63]: Use destructuring:

```
var [secondsFromEpoch,lastSecondFraction] = utc;
```

File: time-ballerina/time_apis.bal [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    decimal lastSecondFraction = utc[1];
+
+    secondsFromEpoch = secondsFromEpoch + <int>seconds.floor();
+    lastSecondFraction = lastSecondFraction + (seconds - seconds.floor());
+    if (lastSecondFraction >= 1) {
+        secondsFromEpoch = secondsFromEpoch + <int>lastSecondFraction.floor();
+        lastSecondFraction = lastSecondFraction - lastSecondFraction.floor();
+    }
+    return <Utc>([secondsFromEpoch, lastSecondFraction].cloneReadOnly());
```
Comment (jclark) [inline time-ballerina/time_apis.bal:72]: Just do:

```
return [secondsFromEpoch,lastSecondFraction];
```

File: time-ballerina/time_apis.bal [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+public isolated function dateValidate(Date date) returns Error? {
+    return externDateValidate(date);
+}
+
+# Get the day of week for a specified date.
+# + date - Date value
+# + return - `DayOfWeek` if the `date` is valid or else panic
+public isolated function dayOfWeek(Date date) returns DayOfWeek {
+    int|Error dayNo = externDayOfWeek(date);
+    if (dayNo is int) {
```
Comment (jclark) [inline time-ballerina/time_apis.bal:97]: Better to lookup in a readonly array.

File: time-ballerina/time_apis.bal [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+# + return - The corresponding `Civil` value
+public isolated function utcToCivil(Utc utc) returns Civil {
+    return externUtcToCivil(utc);
+}
+
+# Converts a given `Civil` value to an `Utc` timestamp.
+# + civilTime - `Civil` time
+# + return - The corresponding `Utc` value or an error if `civilTime.utcOffset` is missing
+public isolated function utcFromCivil(Civil civilTime) returns Utc|Error {
+    if (civilTime["utcOffset"] is ()) {
```
Comment (jclark) [inline time-ballerina/time_apis.bal:136]: Use `?.` operator

File: time-ballerina/time_types.bal [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+public type ZoneOffset readonly & record {|
+                                      int hours;
+                                      int minutes = 0;
+                                      # IETF zone files have historical zones that are offset by
+                                      # integer seconds; we use Seconds type so that this is a subtype
+                                      # of Delta
+                                      decimal seconds?;
+                                  |};
+
+type ReadableZoneOffset record {|
```
Comment (jclark) [inline time-ballerina/time_types.bal:77]: ZoneOffset  is already readable. I suggest `ReadWriteZoneOffset`, but you should not need this.

File: time-ballerina/time_apis.bal (hunk 1 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal:28]: [This language constraint](https://ballerina-platform.slack.com/archives/C9WMA6QDC/p1614848652024000) is the reason for that.

File: time-ballerina/time_types.bal (hunk 8 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_types.bal]: Rename in [here](https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/28f471b68d8d59b493bee7fda2cf9380a4adac85). 

We had to do this due to the [aforementioned language constraint](https://github.com/ballerina-platform/module-ballerina-time/pull/68#discussion_r589200622).

File: time-ballerina/time_apis.bal (hunk 7 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/2934c7c32e78c4ca4ae419b94c9b2e9c0b1751e4

File: time-ballerina/time_apis.bal (hunk 6 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal:85]: We had to do this due to the [aforementioned language constraint](https://github.com/ballerina-platform/module-ballerina-time/pull/68#discussion_r589200622).

File: time-ballerina/time_apis.bal (hunk 5 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/66cfad7ef35768446bf76f504cbe19b76b3a07aa

File: time-ballerina/time_apis.bal (hunk 4 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/66cfad7ef35768446bf76f504cbe19b76b3a07aa

File: time-ballerina/time_apis.bal (hunk 3 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/66cfad7ef35768446bf76f504cbe19b76b3a07aa

File: time-ballerina/time_apis.bal (hunk 2 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/66cfad7ef35768446bf76f504cbe19b76b3a07aa

File: time-ballerina/time_apis.bal [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    secondsFromEpoch = secondsFromEpoch + <int>seconds.floor();
+    lastSecondFraction = lastSecondFraction + (seconds - seconds.floor());
+    if (lastSecondFraction >= 1) {
+        secondsFromEpoch = secondsFromEpoch + <int>lastSecondFraction.floor();
+        lastSecondFraction = lastSecondFraction - lastSecondFraction.floor();
+    }
+    return [secondsFromEpoch, lastSecondFraction];
+}
+
+// TODO Try to do in Ballerina
```
Comment (daneshk) [inline time-ballerina/time_apis.bal:63]:
```suggestion
```

Comment (daneshk) [reply time-ballerina/time_apis.bal:63]: Let's remove this and create an issue to check later

File: time-ballerina/time_errors.bal [hunk 10]
```diff
@@ -14,8 +14,8 @@
...
 
-# Specifies the time error, which occurs in the module.
-public type TimeError distinct error;
+# The error to be returned when parameters are incorrectly formatted .
+public type FormatError distinct error;
 
-# Represents the Time module related error.
-public type Error TimeError;
+# The generic module level error.  
+public type Error FormatError;
```
Comment (daneshk) [inline time-ballerina/time_errors.bal:21]:
```suggestion
public type Error distinct error;
```

File: time-ballerina/time_errors.bal [hunk 11]
```diff
@@ -14,8 +14,8 @@
 // specific language governing permissions and limitations
 // under the License.
 
-# Specifies the time error, which occurs in the module.
-public type TimeError distinct error;
+# The error to be returned when parameters are incorrectly formatted .
+public type FormatError distinct error;
```
Comment (daneshk) [inline time-ballerina/time_errors.bal:18]:
```suggestion
public type FormatError distinct Error;
```

Comment (BuddhiWathsala) [reply time-ballerina/time_errors.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/bf01b736b90ff7099c2fa0aac2f9ff10c9bc8e40

File: time-ballerina/time_errors.bal (hunk 10 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_errors.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/bf01b736b90ff7099c2fa0aac2f9ff10c9bc8e40

File: time-ballerina/time_apis.bal (hunk 9 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/ff05485d52513dd551b26f1471d68394cdfaec03

**Review by daneshk — APPROVED** [review]

File: time-ballerina/time_apis.bal (hunk 6 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_apis.bal:84]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/da9642989fb30f89a20c3afe3318343a835e0933

File: time-ballerina/time_types.bal (hunk 8 above)
Comment (BuddhiWathsala) [reply time-ballerina/time_types.bal]: Fixed in here https://github.com/ballerina-platform/module-ballerina-time/pull/68/commits/da9642989fb30f89a20c3afe3318343a835e0933

**Pull request merged**